File: factory/exporter.py
from abc import ABC, abstractmethod

class VideoExpoter(ABC):
    @abstractmethod
    def prepare_export(self) -> None:
        """To prepafe export"""

    @abstractmethod
    def do_export(self) -> None:
        """Export video"""


class AudioExporter(ABC):
    @abstractmethod
    def convert_audio(self) -> None:
        """convert audio"""


class HightQulatityVideoExporter(VideoExpoter):
    def prepare_export(self) -> None:
        print("hight qulatity video export")

    def do_export(self) -> None:
        print("export hight video")


class LowQulatityVideoExporter(VideoExpoter):
    def prepare_export(self) -> None:
        print("hight low video export")

    def do_export(self) -> None:
        print("export low video")


class Mp3QulatityExporter(AudioExporter):
    def convert_audio(self) -> None:
        print("convert mp3 audio")


class WaveQulatityExporter(AudioExporter):
    def convert_audio(self) -> None:
        print("convert wave audio")


class ExporterFactory(ABC):
    @abstractmethod
    def make_video_exporter(self) -> VideoExpoter:
        """Return video export"""

    @abstractmethod
    def make_audio_exporter(self) -> AudioExporter:
        """Return audio export"""


class HightQulatityExporter(ExporterFactory):
    def make_video_exporter(self) -> VideoExpoter:
        return HightQulatityVideoExporter()

    def make_audio_exporter(self) -> AudioExporter:
        return Mp3QulatityExporter()

class LowQulatityExporter(ExporterFactory):
    def make_audio_exporter(self) -> AudioExporter:
        return WaveQulatityExporter()

    def make_video_exporter(self) -> VideoExpoter:
        return LowQulatityVideoExporter()


def main() -> None:
    qulatities = {
        "high_q": HightQulatityExporter(),
        "low_q": LowQulatityExporter()
    }

    exporter: ExporterFactory

    if (i := "low_q") in qulatities:
        exporter = qulatities[i]

        video_exporter = exporter.make_video_exporter()
        audio_exporter = exporter.make_audio_exporter()

        video_exporter.prepare_export()
        video_exporter.do_export()

        audio_exporter.convert_audio()


if __name__ == "__main__":
    main()

File: factory/exporter.rs
use std::io;

trait VideoExporter {
  fn prepare_export(&self);

  fn do_export(&self);
}


trait AudioExporter {
  fn convert_audio(&self);

  fn do_export(&self);
}


/** For Video Exporter **/
struct HightQulatityVideoExporter;

impl VideoExporter for HightQulatityVideoExporter {
  fn prepare_export(&self) {
    println!("prepare video for hight qulatity");
  }

  fn do_export(&self) {
    println!("exporting video for hight qulatity");
  }
}


struct LowQulatityVideoExporter;

impl VideoExporter for LowQulatityVideoExporter {
  fn prepare_export(&self) {
    println!("prepare video for low qulatity");
  }

  fn do_export(&self) {
    println!("exporting video for low qulatity");
  }
}


/** For Audio Exporter **/
struct Mp3AudioExporter;

impl AudioExporter for Mp3AudioExporter {
  fn convert_audio(&self) {
    println!("Audio convert mp3");
  }

  fn do_export(&self) {
    println!("Audio exporting for Mp3");
  }
}


struct WaveAudioExporter;

impl AudioExporter for WaveAudioExporter {
  fn convert_audio(&self) {
    println!("Audio convert to wave");
  }

  fn do_export(&self) {
    println!("Audio exporting for Wave");
  }
}


/** Export Factory **/
trait ExporterFactory {
  fn make_video_exporter(&self) -> Box<dyn VideoExporter>;

  fn make_audio_exporter(&self) -> Box<dyn AudioExporter>;
}


struct HightQulatityExporter;

impl ExporterFactory for HightQulatityExporter {
  fn make_video_exporter(&self) -> Box<dyn VideoExporter> {
    Box::new(HightQulatityVideoExporter)
  }

  fn make_audio_exporter(&self) -> Box<dyn AudioExporter> {
    Box::new(Mp3AudioExporter)
  }
}


struct LowQulatityExporter;

impl ExporterFactory for LowQulatityExporter {
  fn make_video_exporter(&self) -> Box<dyn VideoExporter> {
    Box::new(LowQulatityVideoExporter)
  }

  fn make_audio_exporter(&self) -> Box<dyn AudioExporter> {
    Box::new(WaveAudioExporter)
  }
}


fn get_exporter() -> Result<Box<dyn ExporterFactory>, io::Error> {
  let mut qulatity = String::new();

  println!("Qulatity:");
  io::stdin().read_line(&mut qulatity)?;

  return match qulatity.trim() {
    "hight" => Ok(Box::new(HightQulatityExporter)),
    "low" => Ok(Box::new(LowQulatityExporter)),
    _ => Err(io::ErrorKind::NotFound.into())
  }
}


fn main() {
  let exporter = get_exporter().expect("Failed exporter");
  
  let video_exporter = exporter.make_video_exporter();
  let audio_exporter = exporter.make_audio_exporter();

  video_exporter.prepare_export();
  video_exporter.do_export();

  audio_exporter.convert_audio();
  audio_exporter.do_export();
}
